File: doublelinkedlist.py


class Node:
    def __init__(self, value):
        self.next = None
        self.prev = None
        self.value = value


class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0
    

    def push(self, value):
        newNode = Node(value)
        if not self.head:
            self.head = newNode;
            self.tail = newNode;
        else:
            self.tail.next = newNode; 
            newNode.prev = self.tail
            self.tail = newNode
        
        self.length += 1
        return self

    def pop(self):
        if not self.head:
            return None
        poppedNode = self.tail
        if self.length == 1:
            self.head = None
            self.tail = None
        else:
            self.tail = poppedNode.prev 
            self.tail.next = None
            poppedNode.prev = None
        self.length -= 1
        return poppedNode

    def shift(self):

        if not self.head:
            return None
        old_head = self.head
        if self.length == 1:
            self.head = None
            self.tail = None
        else: 
            self.head = old_head.next
            old_head.next = None
            old_head.prev = None
        return old_head


    
        


    def print(self):
        current = self.head
        while current is not None:
            if current.value is not None:
                print(current.value)
            current = current.next


x = DoublyLinkedList()
x.push(2)
x.push(3)
x.push(5)
x.print()
x.pop()
x.shift()
print('-----')
x.print()

File: doublelinkedlist.js
class Node{
    constructor(val){
        this.val = val;
        this.next = null;
        this.prev = null;
    }
}

class DoublyLinkedList{
    constructor(){
        this.head = null;
        this.tail = null;
        this.length = 0;
    }
    push(val){
        var newNode = new Node(val)
        if(!this.head){ //if the list is empty
            this.head = newNode;
            this.tail = newNode;
        } else{
            this.tail.next = newNode; // add new node
            newNode.prev = this.tail
            this.tail = newNode; // make it the tail
        }
        this.length++;
        return this;
    }
    pop(){
        if(!this.head){
            return null
        }
       var  poppedNode = this.tail;
       if(this.length == 1){
           this.head = null;
           this.tail = null;
       }else{
          this.tail = poppedNode.prev 
          this.tail.next = null;
          poppedNode.prev = null;
       }
       
       this.length --;
       return poppedNode;
    }
    shift(){
        if(this.length === 0) return undefined;
        var oldHead = this.head;
        if(this.length === 1){
            this.head = null;
            this.tail = null;
        }else{
            this.head = oldHead.next;
            this.head.prev = null;
            oldHead.next = null;
        }
        this.length--;
        return oldHead;
    }
}